ui/app/adapters/kv/metadata.js

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import { encodePath } from 'vault/utils/path-encoding-helpers';
+import ApplicationAdapter from '../application';
+import { kvMetadataPath } from 'vault/utils/kv-path';
+
+export default class KvMetadataAdapter extends ApplicationAdapter {
+  namespace = 'v1';
+
+  _url(fullPath) {
+    return `${this.buildURL()}/${fullPath}`;
+  }
+
+  createRecord(store, type, snapshot) {
+    const { backend, path } = snapshot.record;
+    const id = kvMetadataPath(backend, path);
+    const url = this._url(id);
+    const data = this.serialize(snapshot);
+    return this.ajax(url, 'POST', { data }).then(() => {
+      return {
+        id,
+        data,
+      };
+    });
+  }
+
+  // TODO: replace this with raw request for metadata request?
+  query(store, type, query) {
+    const { backend } = query;
+    return this.ajax(this._url(`${encodePath(backend)}/metadata?list=1`), 'GET').then((resp) => {
+      resp.data.backend = backend;
+      return resp;
+    });
+  }
+
+  queryRecord(store, type, query) {
+    const { backend, path } = query;
+    // ID is the full path for the metadata
+    const id = kvMetadataPath(backend, path);
+    return this.ajax(this._url(id), 'GET').then((resp) => {
+      return {
+        id,
+        ...resp,
+        data: {
+          backend,
+          path,
+          ...resp.data,
+        },
+      };
+    });
+  }
```

Comment:
If I recall from the earlier review there was a need to check the store for the record first to limit requests in a control group scenario. Should there be a `peekRecord` incorporated into this?

Reply:
We decided we'd run peekRecord in the relevant component. See Chelsea's response on her[ PR here](https://github.com/hashicorp/vault/pull/21434/files#diff-c7878795a0aa9b1c7fbcaa68419bf5a15a0996311d65daaa92887bd393cc2b08L7).